presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java

```diff
@@ -1371,20 +1371,22 @@ else if (u21 >= 0) {
         return (int) qhat;
     }
 
-    private static long divideUnsignedLong(long dividend, int divisor)
+    /**
+     * See Hacker's Delight, Section 9.3
+     */
+    private static long divideUnsignedLong(long dividend, long divisor)
     {
-        if (divisor == 1) {
-            return dividend;
+        if (divisor < 0L) {
+            return Long.compareUnsigned(dividend, divisor) < 0 ? 0L : 1L;
         }
-        long unsignedDivisor = divisor & LONG_MASK;
-        long quotient = (dividend >>> 1) / (unsignedDivisor >>> 1);
-        long remainder = dividend - quotient * unsignedDivisor;
-        while (remainder < 0) {
-            remainder += unsignedDivisor;
-            quotient--;
+
+        if (dividend > 0) {
```

Comment:
why is there a special case for dividend > 0?  I haven't spent time thinking through how the algorithm works, but from the chapter you referenced it doesn't look required.

Reply:
This is a shortcircut for the most common case, when both, dividend and divisor are non negative (leftmost bit is 0)